sympy/calculus/util.py

```diff
@@ -1629,3 +1532,133 @@ def union(self, other):
 
 # setting an alias for AccumulationBounds
 AccumBounds = AccumulationBounds
+
+# Hooks
+def Add_hook(*args, **options):
+    evaluate = options.get('evaluate', global_parameters.evaluate)
+    if not evaluate:
+        return None
+
+    extras, accbnds, others = [], [], []
+    for a in args:
+        if isinstance(a, MatrixExpr):
```

Comment:
I really think that we should define the behaviours for the association of pairs of types in associative operators. Single dispatch is probably not enough, I think we need double dispatch.

Reply:
Oops. My bad...

Reply:
Agreed. I'll try it in #19463.

Reply:
Multiple dispatch works when you have a fixed number of types you need to dispatch on e.g. in a binary add `a+b` you only ever need to dispatch on `type(a)` and `type(b)`. In a general add 'a+b+c+...' there can be any number. Multiple dispatch doesn't support variadic functions in general.

Reply:
@oscarbenjamin 
It's true, but I have come to think that with well-designed classes, recursively calling binary dispatcher can solve arbitrary numbered combination.

Previously, in this commit (#18769), this was impossible because `Add` and `Mul` were used as `add` and `mul` constructor, and I had to deal with bunch of back compatibility issues. Instead, please review the recent commit (48b9920) in #19463.